### ISSUE TITLE
False positive `reportMatchNotExhaustive` when matching against a tuple with union type values

### ISSUE THREAD

**Issue by kkom** [microsoft/pyright#8516]
We _think_ that Pyright has some false positive `reportMatchNotExhaustive` errors when matching union types in a tuple.

Take a look at the below code:

```python
def get_value() -> int | str:
    return 1


a = get_value()
b = get_value()

# The below is fine
match a:
    case int():
        pass
    case str():
        pass

# The below triggers a false positive:
#
# > Cases within match statement do not exhaustively handle all values
# >   Unhandled type: "tuple[int | str, int | str]"
# >   If exhaustive handling is not intended, add "case _: pass" Pylance(reportMatchNotExhaustive)
match a, b:
    case int(), int():
        pass
    case int(), str():
        pass
    case str(), int():
        pass
    case str(), str():
        pass


# The below triggers a false positive:
#
# > Cases within match statement do not exhaustively handle all values
# >   Unhandled type: "tuple[int | str, int | str]"
# >   If exhaustive handling is not intended, add "case _: pass" Pylance(reportMatchNotExhaustive)
match a, a:
    case int(), int():
        pass
    case str(), str():
        pass


def get_int() -> int:
    return 1


int_var = get_int()

# The below is fine
match a, int_var:
    case int(), int():
        pass
    case str(), int():
        pass
```

In the second and third example, should Pyright know that the match is exhaustive?

Third example is artificial, but the second one just came up in our codebase.

**VS Code extension or command-line**

We're on `pyright 1.1.372` in the CLI.

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug. 

You're reporting two separate issues above, so let me take each one in turn.

Case 1: `match a, b`:

The initial type of the subject expression is `tuple[int | str, int | str]`. If you match against the pattern `int(), int()`, there's no tuple type that can describe the narrowed fall-through type. To describe this type accurately, a type checker would need to first expand the tuple by creating a union of all possible combinations of tuples. Doing this could result in a combinatoric explosion (O(n!) where n is the number of elements in the tuple type), so care would need to be taken. In your example, the resulting expanded type would be `tuple[int, int] | tuple[int, str] | tuple[str, int] | tuple[str, str]`, and the fall-through narrowed type would be `tuple[int, str] | tuple[str, int] | tuple[str, str]`. This sort of tuple expansion is theoretically possible, but pyright doesn't do it currently.

Pyright will narrow a single element of a tuple during pattern matching, as in the following example. (I'll note that mypy doesn't even handle this case.)

```python
t = a, b
reveal_type(t) # tuple[int | str, int | str]
match t:
    case int(), x:
        reveal_type(t) # tuple[int, int | str]
    case _:
        reveal_type(t) # tuple[str, int | str]
```

So, this case can be considered a feature request for pyright to do full tuple expansion when matching tuples. That's something I've considered doing, but it's a significant amount of work to get right.

I'm going to remove the "bug" tag and change this issue to an enhancement request. I don't plan to implement this feature at this time so I'm going to close the issue, but I would consider revisiting this decision in the future if there is sufficient demand (e.g. in the form of upvotes).

Case 2: `match a, a`:

From the perspective of a type checker, this is effectively the same as case 1. The fact that the tuple was built using the variable `a` twice does not change the fact that the subject tuple has the type `tuple[int | str, int | str]`.

You can see this same effect if you attempt to add `a` to `a`. The expression `a + a` produces a type violation error because `int | str` cannot be added to `int | str`. The fact that you're using the same variable as both operands doesn't matter from the standpoint of a type checker.

So, case 2 is not a bug, nor is it something that's likely to change in the future.

**Comment by erictraut** [comment]
Looking back through pyright's issue tracker, this is actually the fourth issue involving this pattern of tuple type narrowing with a `match` statement, so this appears to be a common pattern. For reference, here are the other three.

https://github.com/microsoft/pyright/issues/4995
https://github.com/microsoft/pyright/issues/6437
https://github.com/microsoft/pyright/issues/7472

I think that's sufficient signal to justify the investment in this feature. Reopening as an enhancement request.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
